src/websockethandler.cpp

```diff
@@ -15,6 +15,60 @@ namespace {
 Logger logger(LOG_MAIN, "WebSocketHandler");
 }  // namespace
 
+ExponentialBackoffStrategy::ExponentialBackoffStrategy() {
+  MVPN_COUNT_CTOR(ExponentialBackoffStrategy);
+
+  m_retryTimer.setSingleShot(true);
+}
+
+#ifdef UNIT_TEST
+void ExponentialBackoffStrategy::testOverrideBaseRetryInterval(
+    int newInterval) {
+  m_baseInterval = newInterval;
+}
+
+void ExponentialBackoffStrategy::testOverrideMaxRetryInterval(int newInterval) {
+  m_maxInterval = newInterval;
+}
+#endif
+
+/**
+ * @brief Schedules the next attempt to execute a given function.
+ *
+ * Everytime a new attempt is scheduled the interval will be exponentially
+ * larger.
+ */
+int ExponentialBackoffStrategy::scheduleNextAttempt(
+    std::function<void()> action) {
+  int retryInterval = qPow(m_baseInterval, m_retryCounter);
+  if (retryInterval < m_maxInterval) {
+    m_retryCounter++;
+  }
+
+  QMetaObject::Connection* const connection = new QMetaObject::Connection;
+  *connection =
+      connect(&m_retryTimer, &QTimer::timeout, this, [action, connection]() {
+        action();
+
+        // Immediatelly disconnect after first execution.
+        QObject::disconnect(*connection);
+        delete connection;
```

Comment:
this connection will be leaked if `scheduleNextAttempt` is called multiple times by mistake.

Reply:
Heh, that is what I get for copying code from stack overflow... This will be gone on my next commit thanks to your previous comment.